### ISSUE TITLE
Breaks `recompile/0` in IEX

### ISSUE THREAD

**Issue by StephenGrider** [AtomLinter/linter-elixirc#80]
**Steps to reproduce**

1) Create a mix project
2) Open Atom with Linter-elixirc running
3) Open the elixir shell via `iex -S mix`
4) Make a change to your project in Atom, save the file.
5) Back at the shell, run `recompile`

**Result**

`recompile` results in a `:noop` indicating that nothing has been recompiled

**Expecation**

`recompile` should recompile the changed modules.

**Thoughts**

`recompile/0` executes `mix compile`, which only recompiles files that have been touched since the last compilation.  I suspect `linter-exlirc` is breaking that process.  Failing a fix, a short term solution may be to add a note to the readme, as this is one TOUGH little problem to debug

**Comment by Arcanemagus** [comment]
It sounds like you want to enable the "Always use elixirc" option which will _significantly_ slow the linting down on large projects. By default this is simply running `mix compile` to only check the changed files. So when you save the changed file this is running `mix compile` and parsing the output, then when you manually run it it sees that nothing has changed since the last run so it does nothing.

tl;dr: This runs `mix compile` already, nothing has changed so it's not broken.

**Comment by StephenGrider** [comment]
I've attached a gif of the issue.  Here's the steps:

1. I successfully run a `main` function in IEX
2. I delete the `main` function in my code editor and save the file.  Presumably linter-elixirc runs.
3. I attempt to `recompile` in IEX, it returns `:noop`. 
4. I again run the `main` function in the shell, and it successfully runs, even though I had deleted the function and recompiled.

![demo5](https://cloud.githubusercontent.com/assets/5003903/19498842/fecf36c2-9547-11e6-9bfd-f9afe3073e87.gif)

Now here's the same process **without** linter-elixirc running.

1. Successfully run `main`.
2. Delete `main`.
3. Run `recompile`.
4. Run `main` again
5. Get an error, as I would expect to see because the function no longer exists.

![demo6](https://cloud.githubusercontent.com/assets/5003903/19498970/ad7596da-9548-11e6-9d41-88994844ec81.gif)

**Comment by StephenGrider** [comment]
To add more clarity - I understand that the linter is recompiling, which is why `recompile/0` does nothing, the undesirable behavior is that it appears the linter's compilation step is breaking iex.

**Comment by Arcanemagus** [comment]
Ah, this goes _way_ beyond my extremely limited understanding of `elixir` then, hopefully somebody in @AtomLinter/linter-elixirc can figure this out.

**Comment by phil-a** [comment]
Having the same issue, and can confirm that disabling Linter-elixirc fixes the recompile returning :noop

**Comment by phoet** [comment]
those things are always the first issues that one runs into when testing a new language or framework 😭 

@StephenGrider nice bootcamp at udemy, i ran into this in lecture 9

**Comment by jon-companyc** [comment]
I have the same issue and when I try to disable the plugin it hangs atom half the time. Tried many times to enable and disable but cant get original recompile functionality back. Any idea how to manually undo what ever this plugin did to recompile in iex?

**Comment by Arcanemagus** [comment]
Okay took some time to try to reproduce this, and I've determined this has _nothing_ to do with this package.

![image](https://cloud.githubusercontent.com/assets/427137/23233850/fccd48fc-f904-11e6-9dd2-1605890fcc0d.png)

As you can see from the above screenshot, the application starts with a `Gh36.foo()` function that simply returns whatever was given to it.

The following steps were ran:

1. Run `Gh36.foo(4)`, observe that it runs properly
2. Delete `GH36.foo()` from the code
3. Run `recompile()` inside IEx, this updates the IEx _in memory_ environment to the current code state.
4. Attempt to run `Gh36.foo(4)` again and observe that it fails, as expected
5. Restore it in the code
6. Run `recompile()` again, again bringing the _in memory_ state of the code up to date
7. Note that the function has been restored and is working
8. Delete the code again
9. **In another terminal** run `mix compile`
10. Attempt to run `recompile()` in IEx, since the "compile state" of the code was updated in the other terminal to say everything is up to date, this does nothing _leaving the in memory state untouched!_
11. Note that although the function is no longer in the code, and `recompile()` thinks things are up to date, the _in memory_ state of the code in IEx is not current.

This is simply a "bug" where `recompile()` inside IEx correctly realizes that the compiled state of the code is up to date, but doesn't recognize that the in-memory version of it doesn't match the current compiled state.

**Comment by phoet** [comment]
> I've determined this has nothing to do with this package

@Arcanemagus i bet my left arm that (maybe at that time) the plugin was the actual issue. whatever side-effect it may have been.

**Comment by mauricerkelly** [comment]
I would say that it is not as clear cut as declaring it a "bug" in IEx. This plugin is lazily relying on the default output directory of elixirc. It is, in effect, contaminating the `build` folder. I can see that `elixirc` has a `-o` option to direct the output files - could this plugin not be adapted to direct the output files somewhere temporary?

(I'm not sure if `mix compile` can be redirected in this way, but at least redirecting the output of `elixirc` is a potential partial solution).

**Comment by Arcanemagus** [comment]
@phoet This is only related to this package by the fact that this compiles the code outside of the IEx instance. _Any_ method that compiles the code outside of that will cause this issue to present itself.

@mauricerkelly The package is already redirecting output to a [temp folder](https://github.com/AtomLinter/linter-elixirc/blob/v1.2.0/lib/init.js#L272-L273), so that isn't the solution unfortunately.

I'm a complete noob with regards to Elixir as I don't use it myself, so I'm open to any suggestions on how to work around this, but as shown above the issue can be reproduced solely from the tools that come with Elixir.

**Comment by mauricerkelly** [comment]
@Arcanemagus i think i had a bad environment for testing this - turned out I was running `linter-elixir-credo` as well, so even though I was forcing `elixirc` compiles in `linter-elixirc`, the other package was still running `mix compile`.

Once I removed `linter-elixir-credo`, forcing `elixirc` in `linter-elixirc` compiles was displaying the behaviour that would be expected.

This still isn't ideal, but I agree that the problem is with `mix` itself -- if `mix` won't allow specifying an alternate output dir on the command line, then we can't really avoid "contamination" of the `_build` directory.

**Comment by mauricerkelly** [comment]
Quick follow-up: one way we could do this is if we used a different environment when we ran `mix` for linting. I wonder if sending `MIX_ENV=test` in the path beforehand would help.

**Comment by speeddragon** [comment]
I think this could be fixed, like @mauricerkelly said. It is confusing some people ... https://github.com/elixir-lang/elixir/issues/6547

**Comment by Arcanemagus** [comment]
@speeddragon That's a workaround for the `elixir` bug, not a fix. This package is only exposing the bug that already exists as outlined in https://github.com/AtomLinter/linter-elixirc/issues/80#issuecomment-281815774.

That being said, that workaround was already applied in #84 and released in v1.3.0 and up, so if you are still seeing this you either haven't updated in a _long_ time, the workaround isn't working any more, or something else is going on.
